### DISCUSSION TITLE
Comparing table (or object) schemas

### DISCUSSION THREAD

**Discussion by sr-murthy** [sqlalchemy/sqlalchemy#6027] [Usage Questions] [answered]
This is a general question about a possible application of SQLAlchemy, and not a bug report.

I don't see any way, currently, of comparing table schemas in a given DB, or between different DB schemas, in SQLAlchemy. 
Is there a way of doing this directly via SQLAlchemy, or would it be something that would be worthwhile to add?

The use case I'm thinking of is one where you have a master DB host and a mirror, where the mirror is updated periodically to reflect the master (via some script). It would be useful, from the testing point of view, to check that all the tables are identical in any given schema between the master and mirror. For example, given SQLAlchemy table objects `table_X` (on master) and `table_X_copy` (on the mirror), it would be nice to something like

    >>> table_X.metadata.schema.equals(table_X_copy.metadata.schema)
    >>>  True

**Comment by zzzeek** [comment, marked as answer]
we have the Alembic feature [compare_metadata](https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html?highlight=compare_metadata#alembic.autogenerate.compare_metadata) for this.   it is oriented towards a reflected database being compared to a fixed MetaData structure .

more generally, if you want to work with two dbs directly, I would advise you write a routine to recursively walk the structures of each, using the [fine grained reflection](https://docs.sqlalchemy.org/en/14/core/reflection.html#fine-grained-reflection-with-inspector) utility.   this will produce simple lists of strings and dicts that can be compared in a straightforward way.

**Reply by sr-murthy** [reply]
Interesting, thanks
